opencog/attentionbank/AttentionBank.cc

```diff
@@ -320,21 +321,23 @@ void AttentionBank::updateAttentionalFocus(const Handle& h,
 
 Handle AttentionBank::getRandomAtomNotInAF(void)
 {
-    std::lock_guard<std::mutex> lock(AFMutex);
-    auto find_in_af = [this](const Handle& h){
-       auto it = std::find_if(attentionalFocus.begin(), attentionalFocus.end(),
-                [h](const std::pair<Handle, AttentionValuePtr>& hsp) { return hsp.first == h; });
-       return it;
-    };
-
-    int count = 50; // We might get stuck in the while loop if there are no atoms outside the AF
-    Handle h = Handle::UNDEFINED;
-    do {
-        h = _importanceIndex.getRandomAtom();
-    } while(find_in_af(h) != attentionalFocus.end() and --count > 0);
-    // Make sure the last selection did not select from the AF.
-    if(find_in_af(h) != attentionalFocus.end())
+    UnorderedHandleSet atoms_not_in_af = _importanceIndex.getHandleSet(0);
+
+    {
+        std::lock_guard<std::mutex> lock(AFMutex);
+        for (const auto& hsp : attentionalFocus) {
+            atoms_not_in_af.erase(hsp.first);
+        }
+    }
+
+    if (atoms_not_in_af.empty())
         return Handle::UNDEFINED;
 
-    return h;
+    auto seed = std::chrono::duration_cast<std::chrono::microseconds>(
+            std::chrono::system_clock::now().time_since_epoch());
```

Comment:
I just copied previous solution here (see https://github.com/opencog/atomspace/blob/0031878db70fa978fb55a18773f7f0f098418c51/opencog/attentionbank/AtomBins.cc#L46)

@nielsgl, my understanding I cannot just use MT19937RandGen::randGen() because it is not thread safe, so user will not be able to use cog-randgen-set-seed to set seed. 

I can write MT19937RandGen rng(0); to keep reproducibility.

Reply:
I don't think you need to worry about setting the random seed. The first time `randGen()` is called, its seed is set to 0 (see https://github.com/opencog/cogutil/blob/master/opencog/util/mt19937ar.cc#L93). It's up to the user to overwrite that seed whenever he/she wants to, from anywhere since this instance in unique.

Yeah, setting the seed isn't probably thread safe, it's never been a problem in practice because it is a seldom operation, but that doesn't mean we should be careless.

Reply:
As far as I see std::mersenne_twister_engine which is base for MT19937RandGen is not thread safe. Method randGen() returns static instance of MT19937RandGen so calling this generator from different threads is not correct. I can use it here but I think we should raise separate issue to synchronize MT19937RandGen.

Reply:
Thank you for advice, will do

Reply:
@linas said
> For the attentionbank, just use the ordinary POSIX C99 rand_r or drand48_r which are thread safe, and are very easy to use.

Wait, but that makes controlling and reproducing the randomness harder, now we'd have 2 sources of randomness in OpenCog processes, 2 seeds to tweak, etc.

What about we only use `randGen()` everywhere?

I know that since C++11 the `RandGen` class has just become a wrapper around `std::mt19937` for backward compatibility + a singleton. But can we instead fix/clean it rather than introducing more sources of randomness?

I'm no expert but it seems to me we have 2 options
1. Make `RandGen` thread safe
2. Use one `RandGen` per thread

I think 1. should be fairly easy to do, to start it seems we don't need to worry about `randGen()`, according to https://stackoverflow.com/questions/2576022/efficient-thread-safe-singleton-in-c#19907903

I don't know which option is better, I'm just not experienced enough with that sort of things, but it seems we need to think about it, maybe create an issue to eventually properly deal with it.

Reply:
The stackoverflow post is about initializing singleton classes, and not about the internal state of that class. The problem is that all random number generators maintain an internal state, and the next random number is a hash of that state. If that state is (for example) a 32-bit int, and two threads simultaneously read that int, then both threads will get exactly the same random number back.  If this happens for the attentionbank, perhaps it doesn't matter. For many monte-carlo apps it does matter.  I'm just fine with ignoring this problem for opencog; it doesn't upset me.

If the state is (for example) a 128-bit int, then it cannot be updated atomically, and one thread might write the lower 64-bits while a different thread writes the upper 64-bits. Again, for opencog, perhaps this does not matter.  Personally, I suspect it does not matter, for us. 

Next. Your 2. is not distinct from 1, it is just one possible solution to 1. The other possible solution is to use a mutex lock in each and every RandGen method.   One randgen-per-thread is almost trivally easy to implement: here's the diff:
```
--- a/opencog/util/mt19937ar.cc
+++ b/opencog/util/mt19937ar.cc
@@ -92,6 +92,6 @@ int MT19937RandGen::rand_discrete(const std::vector<double>& weights)
 // can be changed with the public method RandGen::seed(unsigned long)
 RandGen& opencog::randGen()
 {
-    static MT19937RandGen instance(0);
+    thread_local MT19937RandGen instance(0);
     return instance;
 }
```

Bingo. Now you are thread-safe.  It is also a lot more repeatable, because thread scheduling races no longer affect the next random number. It is also a lot less random, because every thread starts with the same seed -- every thread gets exactly the same random number sequence.

Personally, I think that using `rand_r()` or `drand48_r()` is a superior solution to the above. These would endow the attention bank with it's own private random-number generator, not shared with anyone else.  This gives the best of both worlds: decent random numbers, and reproducibility.

Is there a reason that RandGen a singleton, or is this some bizzaro historical accident dating back to the Brazilian Novamente SpA coding crew? My knee-jerk reaction is that a singleton instance of a random number generator is an insane idea; it provides zero benefits (that I know of, or that I can imagine) while providing a lot of negative downsides.

Reply:
In multithreaded environment sequence of calculations usually is not guaranteed so even synchronized RandGen will not be reproducible. Probably only scenario when RandGen can be predictable when used from multiple threads: 
- thread1 sets seed
- thread1 starts thread2 an join it
- thread2 calls RandGen
Taking this into account I would make RandGen instance thread local as suggested by @linas.

As RandGen is used everywhere I would use it in AttentionBank too. May be rand() is simpler but I don't think it makes a big difference for this method.

@linas, @ngeiswei if you don't mind - I will fix the PR.

Reply:
@vsbogd I agree that reproducibility often requires a single thread anyway. I'm personally OK with making `RandGen` instance thread local.